### PR TITLE
[FIX] website: properly filter website.page records on list view

### PR DESCRIPTION
**Pull request by robinlej**
Since [commit 1], which adapted the website pages list view to OWL, the
records listed on screen are filtered according to the active website
filter. However, the full list of records is still used behind the
scenes for all potential actions.

Steps to reproduce:
1. Go to the pages list view.
2. Select a specific website (if it's not already the case).
=> The total of records in the upper right corner does not match the
number of pages on that specific website.
3. Click on the "Select all" checkbox.
=> All the pages are selected, including those that do not appear on
screen.

This commit updates the records list and kaban when the activeWebsite
changes. Most of the code related to the pages filtering could be
removed after this, but we decided to keep it on the first loading of
the page to avoid a flickering effect (the model is initialized without
the context of existing websites, so it is updated afterwards).
The total of records in the upper right corner is still updated after
the initial load, but is less visible than the whole table being
rerendered.

[commit 1]: https://github.com/odoo/odoo/commit/940f4ee875332dafa1f379970a7683be6b3ee606

opw-3554064
opw-3658648

### CODE REVIEW

**Comment by robinlej** [comment]
@rdeodoo I'm pinging you here since you assigned me.

File: addons/website/static/src/components/views/page_views_mixin.js [hunk 1]
```diff
@@ -21,18 +21,42 @@ export const PageControllerMixin = (component) => class extends component {
...
+                    const domain = [
+                        '|', ['website_id', '=', websiteId],
+                             '&', ['website_id', '=', null],
+                                  ['key', 'not in', activeWebsitePages],
+                    ];
+                    this.env.searchModel.reload({domain});
+                }
+                this.env.searchModel.search();
+            })();
+        },() => [this.state.activeWebsite]);
```
Comment (rdeodoo) [inline addons/website/static/src/components/views/page_views_mixin.js:59]: Why do we need this (query again the DB I mean) if we already have all the records behind the scene available? We can't access those from here?

File: addons/website/static/src/components/views/page_views_mixin.js [hunk 2]
```diff
@@ -84,12 +109,20 @@ export const PageRendererMixin = (component) => class extends component {
...
+        }
         const websiteId = record.data.website_id && record.data.website_id[0];
         return !this.props.activeWebsite.id
             || this.props.activeWebsite.id === websiteId
-            || !websiteId && records.filter(rec => rec.data.website_url === record.data.website_url).length === 1;
+            || !websiteId && !records.some(rec => {
+                    const recWebsiteId = rec.data.website_id && rec.data.website_id[0];
+                    return rec.data.website_url === record.data.website_url
+                            && recWebsiteId === this.props.activeWebsite.id;
+                });
```
Comment (rdeodoo) [inline addons/website/static/src/components/views/page_views_mixin.js:126]: Why do we need to change something here? If I got this right, it's about filtering the list view (and not the counter), which is working already fine, right?
And from what I read in your commit message, this is not related to the counter (which will be wrong briefly), so it's not about that either.

File: addons/website/static/src/components/views/page_views_mixin.js [hunk 3]
```diff
@@ -73,6 +97,7 @@ export const PageControllerMixin = (component) => class extends component {
 
     onSelectWebsite(website) {
         this.state.activeWebsite = website;
+        this.state.firstLoad = false;
```
Comment (rdeodoo) [inline addons/website/static/src/components/views/page_views_mixin.js:100]: Seems like a bandage to hide the issue rather than a real fix imo but I'll wait for your reply on other points before going further and judging this new param

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Isn't there a way to adapt the counter as easily as the way we filter the list view by just patching `recordFilter`?
Actually, checking the code, it looks like we just hid the ones we wanted to hide with `t-if` on the row, maybe there were a more ideal way to do that.

It feels like we kept the full list of pages internally in all the "code engine" of the list view, and then we just hid some pages through simple t-if which is not altering the OWL reference of pages.
Isn't this whole approach wrong and against owl patterns?

I am not familiar enough with it to have a precise answer, we can discuss it tomorrow at the office

File: addons/website/static/src/components/views/page_views_mixin.js (hunk 2 above)
Comment (robinlej) [reply addons/website/static/src/components/views/page_views_mixin.js:126]: Answering this first because it's the easiest and I need to remember the rest:
The filtering is not actually working fine. In case you have a "/contactus" specific to website 1, without the change, if you check the pages of website 2, you won't see the "/contactus" (even though you should have the generic one). Because the previous logic would only consider either pages of website 2, or pages which aren't specific to any website.

File: addons/website/static/src/components/views/page_views_mixin.js (hunk 1 above)
Comment (robinlej) [reply addons/website/static/src/components/views/page_views_mixin.js:59]: From what I understood (but I might be wrong, the ListController and the like are quite intricate, so maybe I overlooked something), we can access the available records, but for all the components (pager, selection, view) to take only the ones we want into consideration, we need to refresh the domain. The whole logic of the controller depends on the search model.
So I'm doing this in a `useEffect()` to update everytime we filter. And I've tried to do it on page load (to avoid the kind of useless `recordFilter()`), but in `onWillStart()`, `this.model.root` isn't set yet, so it's too early, and in the following step it's too late.

I've looked for a better way to do that, but didn't find any. Maybe the filters are tied to an action that should be triggered to update the domain... But if so, which and where?

Comment (robinlej) [reply addons/website/static/src/components/views/page_views_mixin.js:59]: ping @rdeodoo

**Comment by rdeodoo** [comment]
Can you please add opw-3658648 in your commit

File: addons/website/static/src/components/views/page_views_mixin.js (hunk 1 above)
Comment (rdeodoo) [reply addons/website/static/src/components/views/page_views_mixin.js:59]: I tried to see with the frameworkjs how it should have been done (in the first place initially, and now as a fix).
I got some information but no real solution for now, I am waiting for their reply

Comment (rdeodoo) [reply addons/website/static/src/components/views/page_views_mixin.js:59]: Followup: https://github.com/odoo/odoo/pull/148372#issuecomment-1879019780
